cranelift/codegen/src/remove_constant_phis.rs

```diff
@@ -0,0 +1,387 @@
+    }
+    fn set(&mut self, actual: Value, lp: AbstractValue) {
+        match self.absvals.insert(actual, lp) {
+            Some(_old_lp) => {}
+            None => panic!("SolverState::set: formal param {:?} is untracked?!", actual),
+        }
+    }
+}
+
+/// Detect phis in `func` that will only ever produce one value, using a
+/// classic forward dataflow analysis.  Then remove them.
+#[inline(never)]
+pub fn do_remove_constant_phis(func: &mut Function, domtree: &mut DominatorTree) {
+    let _tt = timing::remove_constant_phis();
+    debug_assert!(domtree.is_valid());
+
+    // Get the blocks, in reverse postorder
+    let mut blocks_reverse_postorder = Vec::<Block>::new();
+    for block in domtree.cfg_postorder() {
+        blocks_reverse_postorder.push(*block);
+    }
+    blocks_reverse_postorder.reverse();
+
+    // Phase 1 of 3: for each block, make a summary containing all relevant
+    // info.  The solver will iterate over the summaries, rather than having
+    // to inspect each instruction in each block.
+    let mut summaries = FxHashMap::<Block, BlockSummary>::default();
+
+    for b in &blocks_reverse_postorder {
+        let formals = func.dfg.block_params(*b);
+        let mut summary = BlockSummary::new(SmallVec::from(formals));
+
+        for inst in func.layout.block_insts(*b) {
+            let idetails = &func.dfg[inst];
+            if let BranchInfo::SingleDest(dest, _) = idetails.analyze_branch(&func.dfg.value_lists)
+            {
+                let inst_var_args = func.dfg.inst_variable_args(inst);
+                if inst_var_args.len() > 0 {
+                    let mut actuals = SmallVec::<[Value; 4]>::new();
+                    for arg in inst_var_args {
+                        let arg = func.dfg.resolve_aliases(*arg);
+                        actuals.push(arg);
+                    }
+                    summary.dests.push((inst, dest, actuals));
+                }
+            }
+        }
+
+        summaries.insert(*b, summary);
+    }
+
+    // Phase 2 of 3: iterate over the summaries in reverse postorder,
+    // computing new `AbstractValue`s for each tracked `Value`.  The set of
+    // tracked `Value`s is exactly Group A as described above.
+
+    let entry_block = func
+        .layout
+        .entry_block()
+        .expect("remove_constant_phis: entry block unknown");
+
+    // Set up initial solver state
+    let mut state = SolverState::new();
+
+    for b in &blocks_reverse_postorder {
+        // For each block, get the formals
+        if *b == entry_block {
+            continue;
+        }
+        let formals: &[Value] = func.dfg.block_params(*b);
+        for formal in formals {
+            let mb_old_absval = state.absvals.insert(*formal, AbstractValue::None);
+            assert!(mb_old_absval.is_none());
+        }
+    }
+
+    // Solve: repeatedly traverse the blocks in reverse postorder, until there
+    // are no changes.
+    let mut iter_no = 0;
+    loop {
```

Comment:
It sounds like in practice this converges quickly, but I wonder if it might be worth using a worklist-of-blocks-with-changed-input-state algorithm instead (see e.g. the dataflow analysis in the regalloc checker) to avoid a potential worst case where a small subgraph takes a while to converge within a large function body?

Reply:
For the liveness analysis in the register allocator, I used a worklist initialised in reverse postorder, plus a mechanism to avoid duplicates in the worklist (== all bells and whistles I could think of).  That achieves around 1.8-2.8 evaluations per block on average, for large functions.  From those experiments I saw that the by far the most important factor was to initialise the worklist in reverse postorder, and operate it as a queue thereafter.

This one does something simpler, which is to repeatedly visit the blocks in reverse postorder, until there are no changes.  Many inputs require 3 passes (3.0 evaluations per block), and the worst I saw, for `joey_big.clif`, required 4 passes.  For the following reasons, I concluded that using a worklist wasn't worth the extra complexity:

* given the above numbers, using a worklist wouldn't give a big reduction in the average evals per block

* the worklist itself adds some overhead, and so does having an auxiliary mapping of block to bool, so as to be able to efficiently avoid inserting the same item in the worklist if it is already there

* more than half the expense lies in Phase 1 (building the summaries) and in particular Phase 3 (editing the result); in particular removing formal parameters is expensive because CL's cost for removing just one formal parameter is linear in the number of formals, and that adds up when blocks have 300+ formals.

* for a nonoptimising CLIF->aarch64 translation, this pass costs circa 0.6% in instruction counts, and will be correspondingly less in the opt_level=speed case.

One simple thing I could do is to filter out summaries for blocks that have neither formal parameters nor actual parameters in their jumps, before iterating.  From what I've seen there are many such blocks and they can't possibly contribute anything to the overall result.

Reply:
> One simple thing I could do is to filter out summaries for blocks that have neither formal parameters nor actual parameters in their jumps, before iterating.

I tried that.  For a couple of large tests, it roughly halved the number of summaries (meaning, from around 1/2 the number of blocks to around 1/4 the number of blocks) but reduced the overall insn count for the pass only by around 3%.  And the fixpointing is done over the summaries.  So the fixpointing cost must be pretty minimal already.

I'll leave this refinement in the respin patch.

Reply:
Sounds good -- thanks for the detailed analysis!